bin/crew

```diff
@@ -769,21 +771,23 @@ def prepare_package (destdir)
 
     # check for conflicts with other installed files
     puts "Checking for conflicts with files from installed packages..."
-    conflicts = []
-    @conflictscmd = "grep --exclude #{CREW_META_PATH}#{@pkg.name}.filelist \
-      -Fxf filelist #{CREW_META_PATH}*.filelist | tr ':' ' ' | \
-      sed 's,.filelist,,g' | sed 's,#{CREW_META_PATH},,g'"
-    conflicts << %x[#{@conflictscmd}].chomp.split(" ")
-    conflicts.reject!(&:empty?)
-    unless conflicts.empty?
+    conflicts = {}
+    File.readlines('filelist').each do |p|
+      next unless File.exist?(p.chomp)
+      Dir['/usr/local/etc/crew/meta/*.filelist'].each do |f|
+        pkgName = File.basename(f, '.filelist')
+        c = File.foreach(f).grep( Regexp.new('^' + p + '$') )[0]
+        conflicts[c] = pkgName if c
+      end
```

Comment:
Nice!  Does this add any extra time to the process?

Reply:
Do NOT change this section unless it is faster.

There is a 3 or 4 order of magnitude change in speed from the last version of this code block, and you had better be **damn** sure that this is going to be faster.

Just looking at this code you are iterating a grep for every single file in the file list, which will read the entire list of filelists in CREW_META_PATH for every line in filelist.

On my system that is 19M of files being read for every line in a filelist.
llvm.filelist is ~ 4000 lines.

That's 76Gb of files having to be read from slow ssd just to do a conflict check.

`grep --exclude #{CREW_META_PATH}#{@pkg.name}.filelist \
      -Fxf filelist #{CREW_META_PATH}*.filelist | tr ':' ' ' | \
      sed 's,.filelist,,g' | sed 's,#{CREW_META_PATH},,g'` reads that 19Mb from the filesystem once.

Reply:
It will be faster than `grep` if there is no conflict in the package (because of `File.exist?` wrapper)

Reply:
Let's assume that one has llvm 11 installed, and one is building llvm 12.
Most of those files in llvm12 exist in llvm11, so let's say 4000 of the files in llvm 12 also exist in llvm 11.

https://github.com/skycocker/chromebrew/blob/0013840470a2d4b47438fc7a10ea9ea11ecaca54/bin/crew#L758

So File exists, and we continue onto the code block, where `File.foreach(f).grep( Regexp.new('^' + p + '$') )[0]` has to go through every file from `Dir['/usr/local/etc/crew/meta/*.filelist'].each do |f|`, and that has to happen x the number of files in the llvm 12 filelist which were also in the 11vm 11 filelist.

The total disk read for this algorithm will be... ~ 19Mb x 4000... so the same as before.

Maybe the best method here would be to break these files into little ruby scripts we could test directly and compare?

llvm would be the logical extreme.

`perf stat -r 5 -d <program>` should give us some small examples.

Can we build a test case for this?

Reply:
(I'm building a quick test case.)

Reply:
testcase attached. Also, your change to the algorithm does not seem to work right yet, as it is not showing the actual conflicts, and seems to error out after the first conflict found, which should during comparisons ignore the `@pkg.name.filelist`.
```
perf stat -r 5 -d ruby oldconflicts.rb 
Checking for conflicts with files from installed packages...
Unable to complete this build since there is a conflict with the same file in another package.

gcc10 package conflicts below:
/usr/local/lib64/libgomp.so
Checking for conflicts with files from installed packages...
Unable to complete this build since there is a conflict with the same file in another package.

gcc10 package conflicts below:
/usr/local/lib64/libgomp.so
Checking for conflicts with files from installed packages...
Unable to complete this build since there is a conflict with the same file in another package.

gcc10 package conflicts below:
/usr/local/lib64/libgomp.so
Checking for conflicts with files from installed packages...
Unable to complete this build since there is a conflict with the same file in another package.

gcc10 package conflicts below:
/usr/local/lib64/libgomp.so
Checking for conflicts with files from installed packages...
Unable to complete this build since there is a conflict with the same file in another package.

gcc10 package conflicts below:
/usr/local/lib64/libgomp.so

 Performance counter stats for 'ruby oldconflicts.rb' (5 runs):

            339.06 msec task-clock                #    0.962 CPUs utilized            ( +-  1.97% )
                73      context-switches          #    0.214 K/sec                    ( +-  7.73% )
                 8      cpu-migrations            #    0.024 K/sec                    ( +- 28.96% )
             4,305      page-faults               #    0.013 M/sec                    ( +-  0.05% )
       996,739,450      cycles                    #    2.940 GHz                      ( +-  2.10% )  (50.15%)
     1,351,523,505      instructions              #    1.36  insn per cycle           ( +-  0.25% )  (62.60%)
       237,242,963      branches                  #  699.705 M/sec                    ( +-  0.43% )  (62.50%)
         2,608,311      branch-misses             #    1.10% of all branches          ( +-  2.12% )  (62.66%)
       511,345,096      L1-dcache-loads           # 1508.119 M/sec                    ( +-  0.39% )  (62.72%)
         5,076,733      L1-dcache-load-misses     #    0.99% of all L1-dcache hits    ( +-  2.28% )  (62.87%)
         1,387,491      LLC-loads                 #    4.092 M/sec                    ( +-  4.20% )  (50.36%)
           303,797      LLC-load-misses           #   21.90% of all LL-cache hits     ( +-  7.08% )  (50.24%)

            0.3524 +- 0.0133 seconds time elapsed  ( +-  3.76% )
```

```
perf stat -r 5 -d ruby newconflicts.rb 
Checking for conflicts with files from installed packages...
Unable to complete this build since there is a conflict with the same file in another package.

llvm package conflicts below:
/usr/local/include/llvm/ADT/APFloat.h
Checking for conflicts with files from installed packages...
Unable to complete this build since there is a conflict with the same file in another package.

llvm package conflicts below:
/usr/local/include/llvm/ADT/APFloat.h
Checking for conflicts with files from installed packages...
Unable to complete this build since there is a conflict with the same file in another package.

llvm package conflicts below:
/usr/local/include/llvm/ADT/APFloat.h
Checking for conflicts with files from installed packages...
Unable to complete this build since there is a conflict with the same file in another package.

llvm package conflicts below:
/usr/local/include/llvm/ADT/APFloat.h
Checking for conflicts with files from installed packages...
Unable to complete this build since there is a conflict with the same file in another package.

llvm package conflicts below:
/usr/local/include/llvm/ADT/APFloat.h

 Performance counter stats for 'ruby newconflicts.rb' (5 runs):

            609.27 msec task-clock                #    0.996 CPUs utilized            ( +-  0.32% )
                26      context-switches          #    0.043 K/sec                    ( +- 40.06% )
                 3      cpu-migrations            #    0.005 K/sec                    ( +- 49.74% )
             2,912      page-faults               #    0.005 M/sec                    ( +-  1.97% )
     1,855,294,971      cycles                    #    3.045 GHz                      ( +-  0.34% )  (49.95%)
     2,844,404,546      instructions              #    1.53  insn per cycle           ( +-  0.47% )  (62.45%)
       522,401,396      branches                  #  857.422 M/sec                    ( +-  0.36% )  (62.52%)
         3,166,502      branch-misses             #    0.61% of all branches          ( +-  0.67% )  (62.59%)
     1,000,120,822      L1-dcache-loads           # 1641.508 M/sec                    ( +-  0.24% )  (62.62%)
         6,968,953      L1-dcache-load-misses     #    0.70% of all L1-dcache hits    ( +-  0.72% )  (62.59%)
         2,091,469      LLC-loads                 #    3.433 M/sec                    ( +-  1.13% )  (50.06%)
           348,255      LLC-load-misses           #   16.65% of all LL-cache hits     ( +-  3.49% )  (49.93%)

           0.61168 +- 0.00220 seconds time elapsed  ( +-  0.36% )
```
[testcase.zip](https://github.com/skycocker/chromebrew/files/6337597/testcase.zip)

Reply:
(The `filelist` in that test case is the file list from a llvm12 build.)

Reply:
Separating the check logic into a script...

Reply:
@satmandu I tried to make the logic faster but it is still a bit (~0.2 secs) slower than `grep`, so I restored the original one

Reply:
I wish there was a better way to make it faster... but grep is Really Fast when you don't have to read that stuff over and over again. :)